account_invoice_change_currency/models/account_move.py

```diff
@@ -0,0 +1,93 @@
+        compute="_compute_currency_change_rate",
+        help="Set new currency rate to apply on the invoice.\n"
+        "This rate will be taken in order to convert amounts between the "
+        "currency on the invoice and last currency",
+    )
+    base_currency_id = fields.Many2one(
+        "res.currency",
+        help="Store the base currency when the invoice is created or the"
+        "conversion is called for the first time. "
+        "This is used to calculate conversion from this currency.",
+    )
+    is_base_currency = fields.Boolean(
+        compute="_compute_is_base_currency",
+        help="Check if current currency is the base currency. "
+        "This is used to hide custom rate field in the form view.",
+    )
+
+    @api.model
+    def create(self, values):
+        values.setdefault("base_currency_id", values.get("currency_id"))
+        return super().create(values)
+
+    def action_account_change_currency(self):
+        """
+        This method convert the original price unit from the original
+        currency when the invoice was created to the current currency
+        using the custom rate and recompute all taxes.
+        """
+        today = fields.Date.context_today(self)
+        invoices = self.filtered(lambda x: x.state == "draft").with_context(
+            check_move_validity=False,
+        )
+        for invoice in invoices:
+            if not invoice.base_currency_id:
+                invoice.update({"base_currency_id": invoice.currency_id})
+                invoice.line_ids._set_base_price_unit()
+                continue
+            invoice_date = invoice.invoice_date or today
+            to_currency = invoice.currency_id
+            context = {"custom_rate": invoice.custom_rate, "to_currency": to_currency}
+            original_currency = invoice.base_currency_id.with_context(**context)
+            for line in invoice.line_ids:
+                line.price_unit = original_currency._convert(
+                    line.base_price_unit,
+                    to_currency,
+                    invoice.company_id,
+                    invoice_date,
+                )
+            invoice._recompute_dynamic_lines(recompute_all_taxes=True)
+
+    @api.depends("company_id", "currency_id", "invoice_date")
+    def _compute_currency_change_rate(self):
+        """
+        Compute the custom rate from the original currency when the invoice
+        was created to the current currency. The custom rate field is
+        editable, but it will not change if custom rate is zero or the
+        current currency and the original currency are the same
+        """
+        for invoice in self:
+            if not invoice.currency_id or not invoice.company_id:
+                invoice.custom_rate = 1.0
+                continue
+            date = invoice.invoice_date or fields.Date.context_today(invoice)
+            from_currency = invoice.base_currency_id or invoice.company_currency_id
```

Comment:
Shouldn't it be this way?

```suggestion
            from_currency = invoice.base_currency_id or invoice.currency_id
```

Reply:
company_currency_id was a behavior from v12.0/v13.0, I changed it to currency_id. What do you think about it?